### PR TITLE
Complete Homework 2

### PR DESCRIPTION
**Pull request by yralko**

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+            "objectLiteralDuplicateProperties": false,
+            "objectLiteralShorthandMethods": true,
+            "objectLiteralShorthandProperties": true,
+            "restParams": true,
+            "spread": true,
+            "superInFunctions": true,
+            "templateStrings": true,
+        }
+    },
+    "rules": {
```
Comment (meteorra) [inline .eslintrc]: you can extend plugins like airbnb recommended, typescript recommended etc, not to decline all the rules yourself, just to override any of them

File: .gitignore [hunk 2]
```diff
@@ -1 +1,107 @@
...
+.cache/
+# Comment in the public line in if your project uses Gatsby and not Next.js
+# https://nextjs.org/blog/next-9-1#public-directory-support
+# public
+
+# vuepress build output
+.vuepress/dist
+
+# Serverless directories
+.serverless/
```
Comment (meteorra) [inline .gitignore]: Seems that this file is taken from some project as many directories are not related to current project set up

File: package.json [hunk 3]
```diff
@@ -20,7 +22,25 @@
   },
   "homepage": "https://github.com/yralko/nodejs_gmp#readme",
   "dependencies": {
+    "@hapi/joi": "^16.1.8",
+    "@types/lodash.sortby": "^4.7.6",
+    "@types/uuid": "^3.4.6",
```
Comment (meteorra) [inline package.json]: @types/hapi__joi is missing, project won't compile without it

File: src/task_2.1/index.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+const validator = createValidator()
+
+const app = express();
+
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }))
+
+app.param('id', (req: Request, res: Response, next: NextFunction, value) => {
+
```
Comment (meteorra) [inline src/task_2.1/index.ts]: Did you mean user extraction or anything else in this middleware? What is it for?

File: src/task_2.1/index.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+});
+
+app.get('/user/:id', (req: Request, res: Response, next: NextFunction) => {
+    const { id } = req.params;
+    const user: UserModel = users.find(v => v.id === id && !v.isDeleted);
+
+    if (user) {
+        res.json(user);
+    } else {
+        res.status(400).json(`No user with id ${id} has been found`);
```
Comment (meteorra) [inline src/task_2.1/index.ts]: Resource is absent in the list of resources - 404 error. If not taking it in the context of authorization

File: src/task_2.1/index.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+const app = express();
+
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }))
+
+app.param('id', (req: Request, res: Response, next: NextFunction, value) => {
+    
+    next();
+});
+
```
Comment (meteorra) [inline src/task_2.1/index.ts]: No error middleware?

File: src/task_2.1/index.ts [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    if (user) {
+        res.json(user);
+    } else {
+        res.status(400).json(`No user with id ${id} has been found`);
+    }
+
+    next();
+})
+
+app.post(
```
Comment (meteorra) [inline src/task_2.1/index.ts]: if you are creating REST service, you can use more semantic methods for that. Eg:
"app.route('/products')
        .get(products.list);
        .post(products.create);

app.route('products/:id')
        .get(product.get);
        .delete(product.delete);
        .put(product.update);"

File: src/task_2.1/index.ts [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    const { id } = req.params;
+    const user: UserModel = users.find(v => v.id === id && !v.isDeleted);
+
+    if (user) {
+        res.json(user);
+    } else {
+        res.status(400).json(`No user with id ${id} has been found`);
+    }
+
+    next();
```
Comment (meteorra) [inline src/task_2.1/index.ts]: What is this next for?

File: src/task_2.1/index.ts [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+app.post(
+    '/user/create',
+    validator.body(userCreateValidationSchema),
+    (req: Request, res: Response, next: NextFunction) => {
+        const { login, password, age } = req.body;
+
+        const isLoginExists = users.some(user => user.login === login);
+        
+        if (isLoginExists) {
+            res.json(`User with login ${login} already exists`)
```
Comment (meteorra) [inline src/task_2.1/index.ts]: Should be a  validation middleware with either going next to error middleware or request handler

File: .gitignore (hunk 2 above)
Comment (yralko) [reply .gitignore]: That's correct, this config is taken from https://github.com/github/gitignore

File: package.json (hunk 3 above)
Comment (yralko) [reply package.json]: I have this package in node_modules folder, however it's missing in the dependencies list. Thank you, added.

File: src/task_2.1/index.ts (hunk 4 above)
Comment (yralko) [reply src/task_2.1/index.ts]: Yes, my intention was to extract user Id passed as a parameter

File: src/task_2.1/index.ts (hunk 7 above)
Comment (yralko) [reply src/task_2.1/index.ts]: Thank you, got it

File: src/task_2.1/index.ts (hunk 9 above)
Comment (yralko) [reply src/task_2.1/index.ts]: Thank you, will add

File: src/task_2.1/index.ts (hunk 6 above)
Comment (yralko) [reply src/task_2.1/index.ts]: Thank you, will add

**Pull request merged**